modyn/selector/internal/selector_strategies/downsampling_strategies/rho_loss_downsampling_strategy.py

```diff
@@ -18,11 +25,53 @@ def inform_next_trigger(self, next_trigger_id: int, selector_storage_backend: Ab
         #
         # Will create a new table (pipeline_id, rho_pipeline_id) representing the mapping
         # between pipeline_id and rho_pipeline_id, primary key being pipeline_id.
-        #
+        rho_pipeline_id = self._get_or_create_rho_pipeline_id()
         # Step 2: Prepare the training data for the IL model, by randomly sampling a predefined ratio of samples
         # from next_trigger_id's data and storing them as a TSS with identifier (rho_pipeline_id, next_trigger_id).
-
+        self._prepare_holdout_set(next_trigger_id, rho_pipeline_id, selector_storage_backend)
         # Step 3: Issue training request to the trainer server, with pipeline_id as rho_pipeline_id and trigger_id
         # as next_trigger_id. Wait for the training to complete. Store the model. Record model id in
         # downsampling_params, so that it can be fetched and used for downsampling.
         raise NotImplementedError
+
+    def _get_or_create_rho_pipeline_id(self) -> int:
+        raise NotImplementedError
+
+    def _prepare_holdout_set(
+        self, next_trigger_id: int, rho_pipeline_id: int, selector_storage_backend: AbstractStorageBackend
+    ) -> None:
+        current_trigger_dataset_size = get_trigger_dataset_size(
+            selector_storage_backend, self._pipeline_id, next_trigger_id, tail_triggers=0
+        )
+
+        holdout_set_size = max(int(current_trigger_dataset_size * self.holdout_set_ratio / 100), 1)
+
+        stmt = self._get_holdout_sampling_query(self._pipeline_id, next_trigger_id, holdout_set_size).execution_options(
+            yield_per=self.maximum_keys_in_memory
+        )
+
+        def training_set_producer() -> Iterable[tuple[list[tuple[int, float]], dict[str, Any]]]:
+            with MetadataDatabaseConnection(self._modyn_config) as database:
+                for chunk in database.session.execute(stmt).partitions():
+                    samples = [res[0] for res in chunk]
+                    yield [(sample, 1.0) for sample in samples], {}
+
+        AbstractSelectionStrategy.store_training_set(
```

Comment:
What I think is somehow missing here is marking the samples as used or so in the table. Otherwise, we will be able to have them in the IL set and the actual training set, and I think they need to be exclusive. Or do you plan to do this in a follow up PR?

Reply:
There are several ways to determine the holdout set and the training set for the main model in the paper.
One way is to sample a predefined ratio of the training data as hold out set, and still use the entire training data as training set for the main model. This is how they train on Clothing-1M, and this is what we agreed on last time https://docs.google.com/document/d/19dAm9IfVz4tK8-y8tTuo4lIiy5P7jC-juJveyJYB1AI/edit#heading=h.lnhsfaqrxzvk.

Reply:
Ahh, sorry, I did not realize that " The original training set is used for training the main model" includes these 10%. I checked the RHO-LOSS paper and they indeed seem to do this. Well, I think this might work for big datasets, for stuff like yearbook I don't know, but we can always extend it I guess.